The file contains gene names in rows and sample ID in column. for genes, there are total 174596 rows which include empty rows(undefined gene name) and duplicate gene names. Our objective is to combine multiple rows belonging to same gene and taking its mean. 
the input file: **GSE178967_ZR_NA_col1gene.xlsx** is pre-processed in following ways:
1. Sum of each gene  from all samples with value less than 10 expression are removed
2. gene rows with empty name renamed as NA using bash script.
3. The NA values are not removed in this file but as we wont continue with the NA rows, we will remove the aggregated NA from the output file after aggregate function.




In [ ]:
## input object created for the file GSE178967_ZR_NA_col1gene.xlsx with samples in column and genes are in rows
input <- read.table("/content/GSE178967_ZR_NA_col1gene.txt",header=T,sep="\t") 

In [ ]:
dim(input)


[1] 174596    181

In [ ]:
head(input)

gene       L_136_05 L_173_00 L_112_05 L_282_00  L_133_00  L_150_00  L_210_00 
1 HBB        641550.0 327053.0 934848.0 1495980.0 2305850.0 1338130.0 269498.00
2 MTRNR2L12  234656.0 356521.0 486621.0  334832.0  331533.0  119703.0  20948.80
3 HBA2       207222.0 152607.0 116654.0  344538.0  372509.0  174057.0  41817.30
4 HBA1       118331.0  78276.8 138456.0  222374.0  207864.0  119428.0  32616.60
5 AC010970.1  59824.0 171416.0  52742.0  110868.0   29916.0   19169.0   8053.00
6 PLCG2       61125.5  71365.7  73383.5   93062.4   52323.5   34503.2   7073.39
  L_200_00 L_329_00 ⋯ L_215_05 L_138_05 L_327_00 L_006_00  L_303_00 L_228_00
1 1949460  6077620  ⋯ 1731570  507926   1906840  1274710.0 682778   1452360 
2  360097   712039  ⋯  291622  683470    765457   405682.0 246088    447305 
3  590778  1102000  ⋯  329535  354149   1261350   327047.0 181882    517892 
4  302592   821682  ⋯  251084  571927   1000010   231825.0 131256    292425 
5  205837   560631  ⋯  102411   25960    499182   133197.0 110236    323899 
6  126700   225245  ⋯  115325  239707    917641    89763.5 111230    139194 
  L_127_05 L_270_00 L_012_05  L_170_00 
1 610460.0 622706   1217190.0 1293060.0
2 122628.0 778604    210696.0  199614.0
3 174779.0 162419    140406.0  230477.0
4 110044.0 112814    187562.0  211127.0
5  87622.0 405913     42991.0   82390.0
6  62708.4 345165     94303.7   94080.8

In [ ]:
#agrregate=aggregate(. ~ id, data = input, mean) 
## id is header name of row which have duplicates, please change according to your file column name. here "gene"

agrregate=aggregate(. ~ gene, data = input, mean) 

#convert into a data frame
agrregate<-data.frame(agrregate)


In [ ]:
head(agrregate)

gene    L_136_05  L_173_00   L_112_05   L_282_00  L_133_00  L_150_00 
1 A1BG    48.190000  67.147200  60.425120 36.557020  55.23118  51.47126
2 A1CF     4.418025   5.041700   7.416575  3.544200   3.76895   4.53530
3 A2M     70.250913 178.432592 147.930626 78.076309 108.69165 217.29577
4 A2ML1    8.019005   8.232613   9.788901  5.859114  11.42191  11.27735
5 A2MP1   32.249992  98.999962  75.500068 36.499977  60.50004 121.75003
6 A3GALT2  1.000000   0.000000   4.000000  5.000000   1.00000   0.00000
  L_210_00  L_200_00  L_329_00  ⋯ L_215_05   L_138_05  L_327_00  L_006_00 
1 5.3296400 33.031860 14.160684 ⋯  55.914760 43.015040 19.025384 26.515500
2 0.5029529  3.400800  1.260885 ⋯   3.650300  4.013301  2.058282  2.780300
3 4.5384600 59.499458 21.805698 ⋯ 202.426185 52.529418 84.176595 20.139735
4 0.3762962  8.733236  3.507344 ⋯   5.887839  7.272098  3.732381  6.948733
5 2.5000000 28.500008  6.249998 ⋯ 100.249900 38.499900 46.250108  7.000005
6 1.0000000  7.000000  0.000000 ⋯   6.000000  4.000000  2.000000  1.000000
  L_303_00   L_228_00 L_127_05 L_270_00   L_012_05  L_170_00  
1  35.594680 56.22052 66.05200  47.022000 49.202800  51.505480
2   3.770475  6.65425  4.14415   4.410975  5.168000   4.524801
3 125.352274 60.13796 59.31767 205.144478 49.491946 115.535673
4   8.336410 11.74466 14.79050   7.522965  5.972265   7.544260
5  58.250107 25.99999 35.50002 104.000023 25.000007  48.000080
6   2.000000  1.00000  7.00000   0.000000  8.000000   2.000000

In [ ]:
dim(agrregate)

[1] 34560   181

In [ ]:
write.table(data.frame('ID'=rownames(agrregate), agrregate),file="Aggregate_data.txt",sep='\t',quote = F,row.names = FALSE)


the Aggregate_data.txt is used as input file for PCA and DESEQ2 preprocessing. but before that NA rows need to removed but the output file of aggregate function didnt had NA rows. 
possible reason: Aggregate function might have removed NA automatically or there might be error. so to confirm this, we rerun the code block with the input file in which all NA entries in 1st column (gene) were replaced with NA1. the new input file is: 

In [ ]:
## input object created for the file GSE178967_ZR_NA_col1gene1_NA_1.TXT with samples in column and genes are in rows
input <- read.table("/content/GSE178967_ZR_NA_col1gene1_NA_1.txt",header=T,sep="\t")

In [ ]:
dim(input)

[1] 174596    181

In [ ]:
head(input)

gene       L_136_05 L_173_00 L_112_05 L_282_00  L_133_00  L_150_00  L_210_00 
1 HBB        641550.0 327053.0 934848.0 1495980.0 2305850.0 1338130.0 269498.00
2 MTRNR2L12  234656.0 356521.0 486621.0  334832.0  331533.0  119703.0  20948.80
3 HBA2       207222.0 152607.0 116654.0  344538.0  372509.0  174057.0  41817.30
4 HBA1       118331.0  78276.8 138456.0  222374.0  207864.0  119428.0  32616.60
5 AC010970.1  59824.0 171416.0  52742.0  110868.0   29916.0   19169.0   8053.00
6 PLCG2       61125.5  71365.7  73383.5   93062.4   52323.5   34503.2   7073.39
  L_200_00 L_329_00 ⋯ L_215_05 L_138_05 L_327_00 L_006_00  L_303_00 L_228_00
1 1949460  6077620  ⋯ 1731570  507926   1906840  1274710.0 682778   1452360 
2  360097   712039  ⋯  291622  683470    765457   405682.0 246088    447305 
3  590778  1102000  ⋯  329535  354149   1261350   327047.0 181882    517892 
4  302592   821682  ⋯  251084  571927   1000010   231825.0 131256    292425 
5  205837   560631  ⋯  102411   25960    499182   133197.0 110236    323899 
6  126700   225245  ⋯  115325  239707    917641    89763.5 111230    139194 
  L_127_05 L_270_00 L_012_05  L_170_00 
1 610460.0 622706   1217190.0 1293060.0
2 122628.0 778604    210696.0  199614.0
3 174779.0 162419    140406.0  230477.0
4 110044.0 112814    187562.0  211127.0
5  87622.0 405913     42991.0   82390.0
6  62708.4 345165     94303.7   94080.8

In [ ]:
#agrregate=aggregate(. ~ id, data = input, mean) 
## id is header name of row which have duplicates, please change according to your file column name. here "gene"

aggregate=aggregate(. ~ gene, data = input, mean) 

#convert into a data frame
aggregate<-data.frame(aggregate)

In [ ]:
head(aggregate)

gene    L_136_05  L_173_00   L_112_05   L_282_00  L_133_00  L_150_00 
1 A1BG    48.190000  67.147200  60.425120 36.557020  55.23118  51.47126
2 A1CF     4.418025   5.041700   7.416575  3.544200   3.76895   4.53530
3 A2M     70.250913 178.432592 147.930626 78.076309 108.69165 217.29577
4 A2ML1    8.019005   8.232613   9.788901  5.859114  11.42191  11.27735
5 A2MP1   32.249992  98.999962  75.500068 36.499977  60.50004 121.75003
6 A3GALT2  1.000000   0.000000   4.000000  5.000000   1.00000   0.00000
  L_210_00  L_200_00  L_329_00  ⋯ L_215_05   L_138_05  L_327_00  L_006_00 
1 5.3296400 33.031860 14.160684 ⋯  55.914760 43.015040 19.025384 26.515500
2 0.5029529  3.400800  1.260885 ⋯   3.650300  4.013301  2.058282  2.780300
3 4.5384600 59.499458 21.805698 ⋯ 202.426185 52.529418 84.176595 20.139735
4 0.3762962  8.733236  3.507344 ⋯   5.887839  7.272098  3.732381  6.948733
5 2.5000000 28.500008  6.249998 ⋯ 100.249900 38.499900 46.250108  7.000005
6 1.0000000  7.000000  0.000000 ⋯   6.000000  4.000000  2.000000  1.000000
  L_303_00   L_228_00 L_127_05 L_270_00   L_012_05  L_170_00  
1  35.594680 56.22052 66.05200  47.022000 49.202800  51.505480
2   3.770475  6.65425  4.14415   4.410975  5.168000   4.524801
3 125.352274 60.13796 59.31767 205.144478 49.491946 115.535673
4   8.336410 11.74466 14.79050   7.522965  5.972265   7.544260
5  58.250107 25.99999 35.50002 104.000023 25.000007  48.000080
6   2.000000  1.00000  7.00000   0.000000  8.000000   2.000000

In [ ]:
dim(aggregate)

[1] 34561   181

the dimension of previous formula was 34560 181. here we have 34561 i.e., one additional row for NA1. After downloading now we can remove NA1 and proceed with PCA and DESEQ2. but we will proceed with Aggregate_data.txt file. 

In [ ]:
write.table(data.frame('ID'=rownames(aggregate), aggregate),file="GSE178967_Aggregate_data.txt",sep='\t',quote = F,row.names = FALSE)